### PR TITLE
Fix compatibility breakage from adding NoiseTexture3D

### PR DESCRIPTION
**Pull request by clayjohn**
Follow up to https://github.com/godotengine/godot/pull/76486

https://github.com/godotengine/godot/pull/76486 broke compatibility by introducing a new parameter to ``Noise.get_image()`` and ``Noise.get_seamless_image()``. The new parameter was called ``p_depth`` even though the function did not return a 3D image (i.e. an array of Images), so it really represented a z offset.

This PR introduces a ``get_image_3d()`` for use with 3D textures and it abstracts the old ``get_image()`` into an internal function ``_get_image()`` which handles both 2D and 3D. 

This PR also:
1. Fixes a case of calling wait_on_thread on a thread that hasn't been started
2. Optimizes the generation of normalized noise textures (the speedup is about 33% even for regular 2D textures, it is significantly more for 3D)

Finally, I have left the seamless generation algorithm the same. Although I suspect it can be generalized to the 3D case a bit more and optimized.

### CODE REVIEW

**Review by akien-mga — APPROVED** [review]
Didn't do an in depth review, but the overall changes seem sensible, and undoing the compat breakage is great.

**Comment by Lasuch69** [comment]
This looks awesome! I've wanted to change as little as possible in the current implementation, and now I see that it was unnecessary.

File: modules/noise/doc_classes/Noise.xml [hunk 1]
```diff
@@ -15,12 +15,23 @@
 			<return type="Image" />
 			<param index="0" name="width" type="int" />
 			<param index="1" name="height" type="int" />
+			<param index="2" name="invert" type="bool" default="false" />
+			<param index="3" name="in_3d_space" type="bool" default="false" />
+			<param index="4" name="normalize" type="bool" default="true" />
+			<description>
+				Returns an [Image] containing 2D noise values.
+				Note: With [param normalize] set to [code]false[/code] the default implementation expects the noise generator to return values in the range [code]-1.0[/code] to [code]1.0[/code].
```
Comment (Calinou) [inline modules/noise/doc_classes/Noise.xml:23]:
```suggestion
				[b]Note:[/b] With [param normalize] set to [code]false[/code], the default implementation expects the noise generator to return values in the range [code]-1.0[/code] to [code]1.0[/code].
```

File: modules/noise/doc_classes/Noise.xml [hunk 2]
```diff
@@ -15,12 +15,23 @@
...
+			<param index="1" name="height" type="int" />
 			<param index="2" name="depth" type="int" />
 			<param index="3" name="invert" type="bool" default="false" />
-			<param index="4" name="in_3d_space" type="bool" default="false" />
-			<param index="5" name="normalize" type="bool" default="true" />
+			<param index="4" name="normalize" type="bool" default="true" />
 			<description>
-				Returns a 2D [Image] noise image.
+				Returns an [Array] of [Image]s containing 3D noise values for use with [method ImageTexture3D.create].
 				Note: With [param normalize] set to [code]false[/code] the default implementation expects the noise generator to return values in the range [code]-1.0[/code] to [code]1.0[/code].
```
Comment (Calinou) [inline modules/noise/doc_classes/Noise.xml:35]:
```suggestion
				[b]Note:[/b] With [param normalize] set to [code]false[/code], the default implementation expects the noise generator to return values in the range [code]-1.0[/code] to [code]1.0[/code].
```

File: modules/noise/doc_classes/Noise.xml [hunk 3]
```diff
@@ -66,13 +77,25 @@
 			<return type="Image" />
 			<param index="0" name="width" type="int" />
 			<param index="1" name="height" type="int" />
+			<param index="2" name="invert" type="bool" default="false" />
+			<param index="3" name="in_3d_space" type="bool" default="false" />
+			<param index="4" name="skirt" type="float" default="0.1" />
+			<param index="5" name="normalize" type="bool" default="true" />
+			<description>
+				Returns an [Image] containing seamless 2D noise values.
+				Note: With [param normalize] set to [code]false[/code] the default implementation expects the noise generator to return values in the range [code]-1.0[/code] to [code]1.0[/code].
```
Comment (Calinou) [inline modules/noise/doc_classes/Noise.xml:86]:
```suggestion
				[b]Note:[/b] With [param normalize] set to [code]false[/code], the default implementation expects the noise generator to return values in the range [code]-1.0[/code] to [code]1.0[/code].
```

File: modules/noise/doc_classes/Noise.xml [hunk 4]
```diff
@@ -66,13 +77,25 @@
...
 			<param index="3" name="invert" type="bool" default="false" />
-			<param index="4" name="in_3d_space" type="bool" default="false" />
-			<param index="5" name="skirt" type="float" default="0.1" />
-			<param index="6" name="normalize" type="bool" default="true" />
+			<param index="4" name="skirt" type="float" default="0.1" />
+			<param index="5" name="normalize" type="bool" default="true" />
 			<description>
-				Returns a seamless 2D [Image] noise image.
+				Returns an [Array] of [Image]s containing seamless 3D noise values for use with [method ImageTexture3D.create].
 				Note: With [param normalize] set to [code]false[/code] the default implementation expects the noise generator to return values in the range [code]-1.0[/code] to [code]1.0[/code].
```
Comment (Calinou) [inline modules/noise/doc_classes/Noise.xml:99]:
```suggestion
				[b]Note:[/b] With [param normalize] set to [code]false[/code], the default implementation expects the noise generator to return values in the range [code]-1.0[/code] to [code]1.0[/code].
```

**Review by Calinou — APPROVED** [review]
Tested locally on the MRP linked in https://github.com/godotengine/godot/pull/76486, it works.

Should be good to merge after applying suggestions.

**Comment by clayjohn** [comment]
> Should be good to merge after applying suggestions.

Done!

**Review by Geometror — APPROVED** [review]
Looks great! Unfortunately I haven't had the time to review the original PR, but this definitely looks like a sensible fixup/cleanup. Tested it and it looks like everything is still working fine.

File: modules/noise/noise.h [hunk 5]
```diff
@@ -112,83 +113,140 @@ class Noise : public Resource {
...
-				// Blend s3(Q1) onto s5(Q2) for the top half.
-				T top_blend = _alpha_blend<T>(rd_src(x, y, img_buff<T>::ALT_X), rd_src(x, y, img_buff<T>::DEFAULT), xpos);
-				// Blend s1(Q3) onto Q4 for the bottom half.
-				T bottom_blend = _alpha_blend<T>(rd_src(x, y, img_buff<T>::ALT_XY), rd_src(x, y, img_buff<T>::ALT_Y), xpos);
-				// Blend the top half onto the bottom half.
-				wr(x, y) = _alpha_blend<T>(bottom_blend, top_blend, ypos);
+		// Now blend across z.
+		if (p_depth > 1) {
+			int skirt_depth = MAX(1, p_depth * p_blend_skirt);
+			//int src_depth = p_depth + skirt_depth;
```
[the remark this answers is not in the document]
Comment (clayjohn) [reply modules/noise/noise.h]: Done!

**Comment by Zireael07** [comment]
note that "Depth" and the like are common names for parameters involving z axis

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Thanks for fixing this up! 🎉

**Pull request merged**